### PR TITLE
Feat(#1165) : Collector whale stats

### LINKED ISSUE #1165: Collector Directory
Same as Artist Directory (Spotlight) let's make one laser-focused on collectors
Add identity switch 
- Name of collector
- Number of total bought >0
- Number of total received as gift =0
- Average price of buy per NFT 
- What's distribution of owning NFT per collection
- Total volume made for buys 

| Collector | Bought | Gifted (received) | Avg price of buy | Own NFT per Collection | Volume of buys | 
|-|-|-|-|-|-|

### PR DESCRIPTION
**Pull request by MSghais**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [X] Feature
- [ ] Refactoring

### What's new?

- [X] PR closes #1165
- [X] Collector whale view with column: 
all bought items
hold items
average
volume
scoring holding
30 buy history

### Before submitting Pull Request, please make sure:

- [X] My contribution builds **clean without any errors or warnings**
- [X] I've merged recent default branch -- **main** and I've no conflicts
- [X] I've tried to respect high code quality standards
- [X] I've didn't break any original functionality
- [X] I've posted a screenshot of demonstrated change in this PR

### Optional

- [X] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [X] I've tested PR on mobile and everything seems works
- [X] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [X] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=EFuhGSJpetTgoZ2dfWxwirxADBWgAzWmVfJmTCauZuPxz6z

### Community participation

- [X] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [X] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

image

image

### CODE REVIEW

**Comment by kkukelka** [comment]
deployment preview somehow gets stuck on graphql request for me 🤔 
Screenshot 2022-05-23 at 07 46 40

File: components/collector/SpotlightDetail.vue [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    const {
+      data: { nFTEntities },
+    } = nfts
+
+    this.nfts = nFTEntities?.nodes || []
+  }
+
+  @Watch('account', { immediate: true })
+  watchAccount(val: string, oldVal: string) {
```
Comment (kkukelka) [inline components/collector/SpotlightDetail.vue:50]:
```suggestion
  onAccountChange(val: string, oldVal: string) {
```

File: components/collector/utils.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { Row, SimpleSpotlightNFT } from './types'
+import formatBalance from '@/utils/formatBalance'
+import * as store from '~/store'
+import { getVolume, pairListBuyEvent } from '@/utils/math'
+
+export const nftFn = (a: any): Row => {
```
Comment (kkukelka) [inline components/collector/utils.ts:6]: is there no parameter type for this? ☹️

File: components/collector/utils.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  return {
+    id: a.issuer,
+    total,
+    sold,
+    unique,
+    averagePrice:
+      a.nfts.nodes.filter(onlyOwned).reduce(sumFn, 0) /
+      (a.nfts.nodes.length || 1),
+    count: 1,
+    collectors: 0, // a.nfts.nodes.reduce(uniqueCollectorFn, new Set()),
```
Comment (kkukelka) [inline components/collector/utils.ts:26]: i assume there's no property in the response to calculate this right? can we do a TODO comment here then to solve this in the future

File: components/collector/utils.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import { Row, SimpleSpotlightNFT } from './types'
+import formatBalance from '@/utils/formatBalance'
+import * as store from '~/store'
+import { getVolume, pairListBuyEvent } from '@/utils/math'
+
+export const nftFn = (a: any): Row => {
+  const sold = a.nfts.nodes.reduce(soldFn, 0)
+  const unique = a.nfts.nodes.reduce(uniqueFn, new Set()).size
```
Comment (kkukelka) [inline components/collector/utils.ts:8]:
```suggestion
  const sold = a.nfts.nodes?.reduce(soldFn, 0)
  const unique = a.nfts.nodes?.reduce(uniqueFn, new Set())?.size
```

File: components/collector/utils.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  const total = a.nfts.totalCount
+  const uniqueCollectors = a.nfts.nodes.reduce(
+    uniqueCollectorFn,
+    new Set()
+  ).size
+  const buyEvents = a.nfts.nodes.map(onlyEvents).map(pairListBuyEvent).flat()
```
Comment (kkukelka) [inline components/collector/utils.ts:14]:
```suggestion
  const uniqueCollectors = a.nfts.nodes?.reduce(
    uniqueCollectorFn,
    new Set()
  )?.size
  const buyEvents = a.nfts.nodes?.map(onlyEvents)?.map(pairListBuyEvent)?.flat()
```

File: pages/collector-whale.vue [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    Identity: () => import('@/components/shared/format/Identity.vue'),
+    CollectorTable: () => import('@/components/collector/CollectorTable.vue'),
+  },
+  head() {
+    const title = 'NFT Collector rank'
+    const metaData = {
+      title,
+      type: 'profile',
+      description: 'Discover new collectors based on ranking',
+      url: '/spotlight',
```
Comment (kkukelka) [inline pages/collector-whale.vue:33]:
```suggestion
      url: '/collector-whale',
```

File: pages/collector-whale.vue [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  components: {
+    Identity: () => import('@/components/shared/format/Identity.vue'),
+    CollectorTable: () => import('@/components/collector/CollectorTable.vue'),
+  },
+  head() {
+    const title = 'NFT Collector rank'
+    const metaData = {
+      title,
+      type: 'profile',
+      description: 'Discover new collectors based on ranking',
```
Comment (kkukelka) [inline pages/collector-whale.vue:32]:
```suggestion
      description: 'Discover the most active collector whales on RMRK.',
```

File: pages/collector-whale.vue [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+<script lang="ts">
+import { Component, Vue } from 'nuxt-property-decorator'
+
+@Component<CollectorWhale>({
+  components: {
+    Identity: () => import('@/components/shared/format/Identity.vue'),
+    CollectorTable: () => import('@/components/collector/CollectorTable.vue'),
+  },
+  head() {
+    const title = 'NFT Collector rank'
```
Comment (kkukelka) [inline pages/collector-whale.vue:28]:
```suggestion
    const title = 'Collector Whales on RMRK'
```

File: langDir/en.json [hunk 9]
```diff
@@ -546,6 +546,21 @@
     "uniqueCollectors": "∃! collectors",
     "soldHistory": "Last 30 Days Sold"
   },
+  "collector": {
+    "page": "Collector",
+    "title": "Collector Whale",
+    "subtitle": "Featuring the most active collectors whale on RMRK.",
```
Comment (kkukelka) [inline langDir/en.json:552]:
```suggestion
    "subtitle": "Featuring the most active collector whales on RMRK.",
```

**Review by kkukelka — CHANGES_REQUESTED** [review]

**Review by kkukelka — APPROVED** [review]
lgtm, waiting for new rubick deploy

**Comment by MSghais** [comment]
Hi @vikiival, this PR is tested. 
Could you merge the collector_whale on the rubick server, and deployed it please, for this PR please ?

**Comment by vikiival** [comment]
> Hi @vikiival, this PR is tested.

Lovely to hear waiting for the rubick testing

File: components/collector/CollectorDetail.vue [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+@Component({ components })
+export default class CollectorDetail extends mixins(PrefixMixin) {
+  @Prop(String) public account!: string
+  protected nfts: NftSimpleView[] = []
+  protected isLoading = true
+
+  protected async fetchNFT(account: string) {
+    const nfts = await this.$apollo.query({
+      query: nftSimpleListByAccount,
+      client: 'rmrk',
```
Comment (vikiival) [inline components/collector/CollectorDetail.vue:34]: why is `rmrk` here?

**Review by vikiival — CHANGES_REQUESTED** [review]
Does this work? 
you use `rmrk` indexer but queries are for subsquid?

File: components/collector/CollectorDetail.vue [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+@Component({ components })
+export default class CollectorDetail extends mixins(PrefixMixin) {
+  @Prop(String) public account!: string
+  protected nfts: NftSimpleView[] = []
+  protected isLoading = true
+
+  protected async fetchNFT(account: string) {
+    const nfts = await this.$apollo.query({
+      query: nftSimpleListByAccount,
+      client: 'subsquid',
```
Comment (vikiival) [inline components/collector/CollectorDetail.vue:34]:
```suggestion
      client: this.client,
```

File: components/collector/CollectorDetail.vue [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      <GalleryCard :id="nft.id" :name="nft.name" :metadata="nft.metadata" />
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+import { Component, Prop, mixins, Watch } from 'nuxt-property-decorator'
+import shouldUpdate from '@/utils/shouldUpdate'
+import PrefixMixin from '@/utils/mixins/prefixMixin'
+import nftSimpleListByAccount from '@/queries/nftSimpleListByAccount.graphql'
```
Comment (vikiival) [inline components/collector/CollectorDetail.vue:13]: pls use correct path

**Review by vikiival — CHANGES_REQUESTED** [review]
why the code has like 400 LOC?

**Comment by vikiival** [comment]
Pls resolve conflicts, fix the suggestions 

and we can merge

**Comment by kkukelka** [comment]
> why the code has like 400 LOC?

CollectorTable.vue takes up a chunk of it, which seems valid though

**Comment by vikiival** [comment]
Marked as unread ✅

**Comment by petersopko** [comment]
@MSghais any updates on this?

File: components/collector/CollectorDetail.vue [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+      },
+      fetchPolicy: 'network-only',
+    })
+
+    const {
+      data: { nFTEntities },
+    } = nfts
+
+    this.nfts = nFTEntities?.nodes || []
+  }
```
Comment (vikiival) [inline components/collector/CollectorDetail.vue:47]: I doubt that this works, please make it workable for all chains

**Review by vikiival — CHANGES_REQUESTED** [review]

**Comment by petersopko** [comment]
@MSghais mate, I'm sorry but we'll have to close this one. We're happy that you're contributing but this particular PR is here for a very long time, without improvement. Some feedback so we can avoid wasting your time in the future:
- if you're not sure about the codebase, I'd suggest picking up smaller issues. There are many issues connected to basilisk which are fairly easy to implement, I expect that after the indexer for moonsama is done, there will be many more.
- if you open up a PR, you should see the end of the task you're solving around the corner
- if you're not sure about the finishing of the task, you can open up a PR with unfinished code and expect help, but in that case, **your communication is required**
- try to keep the communication in the open (comments/reviews). I see that you're sending me DMs about the PRs you're solving, I'm, generally speaking, avoiding replying to those as we're working here as a team and there are multiple people participating :)

At time of closing, deploy has failed some time ago, all the checks are not passing and you're not responding to review comments for ~4 days. Feel free to reopen with new code, or ask the questions you have in the issue you're trying to solve.

**Pull request closed without merging**